Refactor to follow Fractal codebase

neilsimp1 opened the pull request:
Hi Johan,

I contacted you some time ago about this. After that, I jumped into this and got in over my head too quickly and lost interest in this.

A couple weeks ago I finally got it in me to look into this app again and I thought about what you were saying about Fractal being a good model to follow code-wise.
So, I took some time and basically re-wrote what you had to mostly follow how Fractal is structured. I used a lot of your original code and a lot from Fractal directly.
That said, I am still really new to Rust and GTK so this may not be perfect. But, I think it has feature parity with what you built.

So, let me know what you think. I have had fun doing this and want to keep adding features, and I have a ton of ideas, so I will keep hacking away at this.

Oh, and to your email, if you accept this PR, I too would be in favor of GPL-ing this program.

Cheers!

neilsimp1 committed:
Refactor to follow Fractal codebase

neilsimp1 committed:
Update README.md

neilsimp1 committed:
Restructureed comments widget and view and css

neilsimp1 committed:
Add collapsing comments

johan-bjareholt commented:
Awesome, really appreciate these changes! :)

I'm busy at the moment, will take a look at it during the weekend.

johan-bjareholt commented on src/appop/subreddit.rs:
```diff
@@ -0,0 +1,79 @@
+impl AppOp {
+    pub fn load_subreddit_view(&self, subreddit_name: String) {        
+        let mut client = Client::new();
+        let posts = client.get_subreddit_posts(&subreddit_name);
```
Making synchronous HTTP requests on the main thread freezes and makes the the application unresponsive while loading.
We should rather make the request in the "backend" thread and then pass the posts to this function running on the main thread. Shouldn't be to hard to fix.

johan-bjareholt commented on src/main.rs:
```diff
@@ -1,353 +1,37 @@
-        let ctx = glib::MainContext::default();
-        let mut prev_view_stack : LinkedList<ViewChangeCommand> = LinkedList::new();
-        loop {
-            set_loadingspinner(false);
```
Removal of the spinner is a shame.
Well, but when the HTTP requests were running on the main thread I guess it's now broken anyway, but should be easy to fix!

johan-bjareholt commented on src/appop/comments.rs:
```diff
@@ -0,0 +1,70 @@
+    pub fn load_comments_view(&self, post_id: String, subreddit_name: String) {let mut client = Client::new();
+        let comment_list = client.get_comments(&post_id).unwrap();
+
+        let comments_headerbar = self.create_comments_headerbar(&subreddit_name);
```
Kind of strange to have one headerbar for each view, would make more sense to have a "GtkStack" for the sub-view (such as subreddit, comments, webview) and then keep the same header but just changing the title/subtitle.

johan-bjareholt reviewed (approved):
Huge improvements, thanks a lot for this!

Didn't take a deep look at all the changes but found a couple major regressions, but the improvements are vastly more valuable so I'll merge this and then you or I can fix those afterwards.

neilsimp1 commented:
I did notice the app getting "stuck" when waiting for HTTP requests, but wasn't really sure how to fix it. I managed to hack most of what I did by copy/pasting and there's a lot of things I didn't understand fully around this. So I do really appreciate the feedback.

Regarding the loading spinner - whoops, I planned on implementing that back in in a slightly different way (show in the main_window instead of headerbar since BaconReader on Android, which I use, has it that way), but I forgot that part. I'm not strongly in favor of it going in the headerbar or main_window or wherever but I do think it's good to have so the user knows it's not just frozen up.

The headerbar - I'm not very experienced with GTK so any improvements if there are better ways of doing things I am open. The reason I changed them up between views was again due to me looking at BaconReader as a model, and again it doesn't necessarily need to go that direction.

Since this PR, I have made one update to my fork - comments are collapsible/expandable when you tap on them. It's not perfect but I think it looks kind of nice so far.
In doing this, I was talking with someone on the Libhandy matrix channel who recommended that instead of having `gtk:Box`s within `gtk:Box`s, (such as the lists of comment boxes or list of post boxes) to instead use `gtk::ListBox` and `gtk::ListBoxRow`. Even for the tree comment structure, he said this would be a good approach, and pointed me to [gitg](https://gitlab.gnome.org/GNOME/gitg/-/blob/master/gitg/history/gitg-history-refs-list.vala) as an example. So, this is why I switched the Comments view structure up so much in the last commit of my fork.
I was planning on doing the Posts view this way too, since `gtk::ListBox` has handy methods like sorting/filtering.


I may have time to look in to some of this soon.
I'm also thinking of a couple of quick feature additions for easy wins, such as a "Open in browser" button on Web views, but what you've pointed out is more important, especially getting requests out of the UI thread.
